solutions/src/main/java/ru/sberbank/school/task02/ExtendedCalculator.java

```diff
@@ -0,0 +1,83 @@
+package ru.sberbank.school.task02;
+
+import java.math.BigDecimal;
+import java.math.MathContext;
+import java.util.*;
+
+import ru.sberbank.school.task02.util.*;
+
+/**
+ * Реализация обратного валютного калькулятора.
+ */
+public class ExtendedCalculator extends Calculator implements ExtendedFxConversionService {
+
+    public ExtendedCalculator(ExternalQuotesService externalQuotesService) {
+        super(externalQuotesService);
+    }
+
+    @Override
+    public Optional<BigDecimal> convertReversed(ClientOperation operation, Symbol symbol, BigDecimal amount,
+                                                Beneficiary beneficiary) {
+
+        List<Quote> quotes = externalQuotesService.getQuotes(symbol);
+        if (quotes == null || quotes.isEmpty()) {
+            throw new EmptyQuoteException("No quotes!");
+        }
+
+        List<ReverseQuote> reverseQuotes = getReverseQuotes(quotes, operation);
+        List<BigDecimal> prices = new ArrayList<>();
+
+        for (ReverseQuote r : reverseQuotes) {
+            if ((r.getVolumeFrom().isInfinity() || amount.compareTo(r.getVolumeFromSize()) >= 0)
+                     && (r.getVolumeTo().isInfinity() || amount.compareTo(r.getVolumeToSize()) < 0)) {
+                prices.add(r.getPrice());
+            }
+        }
+
+        if (prices.isEmpty()) {
+            return Optional.empty();
+        }
+
+        Collections.sort(prices);
+        return Optional.of(beneficiary == Beneficiary.CLIENT ^ operation == ClientOperation.BUY ? prices.get(prices.size() - 1) : prices.get(0));
```

Comment:
Line is longer than 120 characters (found 145).

*<this is a auto generated comment from violation-comments-lib F7F8ASD8123FSDF>* *<a-1559895052>*